flask requires `is_package` from import loaders, however this api is optional since PEP 451

For more information: https://github.com/pytest-dev/pytest/issues/5532

https://www.python.org/dev/peps/pep-0451/#api-changes

A compliant PEP 451 loader can avoid a lot of the tricky APIs just through implementing `find_spect` -- in this case `is_package` is replaced by `submodule_search_locations` being truthy

I'll see if I can't take a stab at this, it looks like this is a reopen of #487

Here's a minimal example:

```python
import flask

app = flask.Flask(__name__)

def test():
    pass
```

```console
$ pytest t.py
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-5.0.0, py-1.8.0, pluggy-0.12.0
rootdir: /tmp/x
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
____________________________ ERROR collecting t.py _____________________________
t.py:3: in <module>
    app = flask.Flask(__name__)
venv/lib/python3.6/site-packages/flask/app.py:381: in __init__
    instance_path = self.auto_find_instance_path()
venv/lib/python3.6/site-packages/flask/app.py:678: in auto_find_instance_path
    prefix, package_path = find_package(self.import_name)
venv/lib/python3.6/site-packages/flask/helpers.py:826: in find_package
    loader, root_mod_name):
venv/lib/python3.6/site-packages/flask/helpers.py:789: in _matching_loader_thinks_module_is_package
    loader.__class__.__name__)
E   AttributeError: AssertionRewritingHook.is_package() method is missing but is required by Flask of PEP 302 import hooks.  If you do not use import hooks and you encounter this error please file a bug against Flask.
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.31 seconds ============================
```

```console
$ pip freeze --all
atomicwrites==1.3.0
attrs==19.1.0
Click==7.0
Flask==1.0.3
importlib-metadata==0.18
itsdangerous==1.1.0
Jinja2==2.10.1
MarkupSafe==1.1.1
more-itertools==7.1.0
packaging==19.0
pip==19.1.1
pluggy==0.12.0
py==1.8.0
pyparsing==2.4.0
pytest==5.0.0
setuptools==41.0.1
six==1.12.0
wcwidth==0.1.7
Werkzeug==0.15.4
wheel==0.33.4
zipp==0.5.1
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -10,6 +10,7 @@ Unreleased
     requires upgrading to Werkzeug 0.15.5. :issue:`3249`
 -   ``send_file`` url quotes the ":" and "/" characters for more
     compatible UTF-8 filename support in some browsers. :issue:`3074`
+-   Fixes for PEP451 import loaders and pytest 5.x.  :issue:`3275`
 
 
 Version 1.0.3

--- a/docs/tutorial/tests.rst
+++ b/docs/tutorial/tests.rst
@@ -188,7 +188,7 @@ should be closed.
         with pytest.raises(sqlite3.ProgrammingError) as e:
             db.execute('SELECT 1')
 
-        assert 'closed' in str(e)
+        assert 'closed' in str(e.value)
 
 The ``init-db`` command should call the ``init_db`` function and output
 a message.

--- a/examples/tutorial/tests/test_db.py
+++ b/examples/tutorial/tests/test_db.py
@@ -12,7 +12,7 @@ def test_get_close_db(app):
     with pytest.raises(sqlite3.ProgrammingError) as e:
         db.execute('SELECT 1')
 
-    assert 'closed' in str(e)
+    assert 'closed' in str(e.value)
 
 
 def test_init_db_command(runner, monkeypatch):

--- a/flask/helpers.py
+++ b/flask/helpers.py
@@ -789,19 +789,38 @@ def _matching_loader_thinks_module_is_package(loader, mod_name):
         loader.__class__.__name__)
 
 
-def find_package(import_name):
-    """Finds a package and returns the prefix (or None if the package is
-    not installed) as well as the folder that contains the package or
-    module as a tuple.  The package path returned is the module that would
-    have to be added to the pythonpath in order to make it possible to
-    import the module.  The prefix is the path below which a UNIX like
-    folder structure exists (lib, share etc.).
-    """
-    root_mod_name = import_name.split('.')[0]
+def _find_package_path(root_mod_name):
+    """Find the path where the module's root exists in"""
+    if sys.version_info >= (3, 4):
+        import importlib.util
+
+        try:
+            spec = importlib.util.find_spec(root_mod_name)
+            if spec is None:
+                raise ValueError("not found")
+        # ImportError: the machinery told us it does not exist
+        # ValueError:
+        #    - the module name was invalid
+        #    - the module name is __main__
+        #    - *we* raised `ValueError` due to `spec` being `None`
+        except (ImportError, ValueError):
+            pass  # handled below
+        else:
+            # namespace package
+            if spec.origin in {"namespace", None}:
+                return os.path.dirname(next(iter(spec.submodule_search_locations)))
+            # a package (with __init__.py)
+            elif spec.submodule_search_locations:
+                return os.path.dirname(os.path.dirname(spec.origin))
+            # just a normal module
+            else:
+                return os.path.dirname(spec.origin)
+
+    # we were unable to find the `package_path` using PEP 451 loaders
     loader = pkgutil.get_loader(root_mod_name)
-    if loader is None or import_name == '__main__':
+    if loader is None or root_mod_name == '__main__':
         # import name is not found, or interactive/main module
-        package_path = os.getcwd()
+        return os.getcwd()
     else:
         # For .egg, zipimporter does not have get_filename until Python 2.7.
         if hasattr(loader, 'get_filename'):
@@ -815,17 +834,29 @@ def find_package(import_name):
             # Google App Engine's HardenedModulesHook
             #
             # Fall back to imports.
-            __import__(import_name)
-            filename = sys.modules[import_name].__file__
+            __import__(root_mod_name)
+            filename = sys.modules[root_mod_name].__file__
         package_path = os.path.abspath(os.path.dirname(filename))
 
         # In case the root module is a package we need to chop of the
         # rightmost part.  This needs to go through a helper function
         # because of python 3.3 namespace packages.
-        if _matching_loader_thinks_module_is_package(
-                loader, root_mod_name):
+        if _matching_loader_thinks_module_is_package(loader, root_mod_name):
             package_path = os.path.dirname(package_path)
 
+    return package_path
+
+
+def find_package(import_name):
+    """Finds a package and returns the prefix (or None if the package is
+    not installed) as well as the folder that contains the package or
+    module as a tuple.  The package path returned is the module that would
+    have to be added to the pythonpath in order to make it possible to
+    import the module.  The prefix is the path below which a UNIX like
+    folder structure exists (lib, share etc.).
+    """
+    root_mod_name, _, _ = import_name.partition('.')
+    package_path = _find_package_path(root_mod_name)
     site_parent, site_folder = os.path.split(package_path)
     py_prefix = os.path.abspath(sys.prefix)
     if package_path.startswith(py_prefix):

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -1219,17 +1219,17 @@ def test_response_type_errors():
 
     with pytest.raises(TypeError) as e:
         c.get('/none')
-        assert 'returned None' in str(e)
+        assert 'returned None' in str(e.value)
 
     with pytest.raises(TypeError) as e:
         c.get('/small_tuple')
-        assert 'tuple must have the form' in str(e)
+        assert 'tuple must have the form' in str(e.value)
 
     pytest.raises(TypeError, c.get, '/large_tuple')
 
     with pytest.raises(TypeError) as e:
         c.get('/bad_type')
-        assert 'it was a bool' in str(e)
+        assert 'it was a bool' in str(e.value)
 
     pytest.raises(TypeError, c.get, '/bad_wsgi')
 
@@ -1622,7 +1622,7 @@ def test_debug_mode_complains_after_first_request(app, client):
         @app.route('/foo')
         def broken():
             return 'Meh'
-    assert 'A setup function was called' in str(e)
+    assert 'A setup function was called' in str(e.value)
 
     app.debug = False
 
@@ -1677,9 +1677,9 @@ def test_routing_redirect_debugging(app, client):
     with client:
         with pytest.raises(AssertionError) as e:
             client.post('/foo', data={})
-        assert 'http://localhost/foo/' in str(e)
+        assert 'http://localhost/foo/' in str(e.value)
         assert ('Make sure to directly send '
-                'your POST-request to this URL') in str(e)
+                'your POST-request to this URL') in str(e.value)
 
         rv = client.get('/foo', data={}, follow_redirects=True)
         assert rv.data == b'success'

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -468,8 +468,8 @@ class TestSendfile(object):
     def test_send_file_object_without_mimetype(self, app, req_ctx):
         with pytest.raises(ValueError) as excinfo:
             flask.send_file(StringIO("LOL"))
-        assert 'Unable to infer MIME-type' in str(excinfo)
-        assert 'no filename is available' in str(excinfo)
+        assert 'Unable to infer MIME-type' in str(excinfo.value)
+        assert 'no filename is available' in str(excinfo.value)
 
         flask.send_file(StringIO("LOL"), attachment_filename='filename')
 

